apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,281 +277,181 @@ terminate(_Reason, _State) ->
+	#mining_candidate{ h1 = H1 } = Candidate,
 	State2 = hash_computed(h1, Candidate, State),
-	case h1_passes_diff_checks(H1, Candidate, State2) of
+	State3 = case h1_passes_diff_checks(H1, Candidate, State2) of
 		true ->
+			%% h1 solution found, report it.
 			?LOG_INFO([{event, found_h1_solution},
 				{step, Candidate#mining_candidate.step_number},
 				{worker, State2#state.name},
-				{h1, ar_util:encode(H1)}, 
+				{h1, ar_util:encode(H1)},
 				{p1, Candidate#mining_candidate.partition_number},
 				{difficulty, get_difficulty(State2, Candidate)}]),
-			ar_mining_stats:h1_solution(),
-			%% Decrement 1 for chunk1:
-			%% Since we found a solution we won't need chunk2 (and it will be evicted if
-			%% necessary below)
-			State3 = remove_sub_chunks_from_cache(Candidate, 1, State2),
 			ar_mining_server:prepare_and_post_solution(Candidate),
-			{noreply, State3};
-		Result ->
-			case Result of
-				partial ->
-					ar_mining_server:prepare_and_post_solution(Candidate);
-				_ ->
-					ok
-			end,
-			{ok, Config} = application:get_env(arweave, config),
-			case cycle_sub_chunk_cache(Candidate, {chunk1, Chunk1, H1}, State2) of
-				{cached, State3} ->
-					%% Chunk2 hasn't been read yet, so we cache Chunk1 and wait for
-					%% Chunk2 to be read.
-					{noreply, State3};
-				{do_not_cache, State3} ->
-					%% This node does not store Chunk2. If we're part of a coordinated
-					%% mining set, we can try one of our peers, otherwise we're done.
-					case Config#config.coordinated_mining of
-						false ->
-							ok;
-						true ->
-							DiffPair =
-								case get_partial_difficulty(State3, Candidate) of
-									not_set ->
-										get_difficulty(State3, Candidate);
-									PartialDiffPair ->
-										PartialDiffPair
-								end,
-							ar_coordination:computed_h1(Candidate, DiffPair)
-					end,
-					%% Decrement 1 for chunk1:
-					%% do_not_cache indicates chunk2 was not and will not be read or cached
-					{noreply, update_sub_chunk_cache_size(-1, SessionKey, State3)};
-				{{chunk2, Chunk2}, State3} ->
-					%% Chunk2 has already been read, so we can compute H2 now.
-					ar_mining_hash:compute_h2(
-						self(), Candidate#mining_candidate{ chunk2 = Chunk2 }),
-					%% Decrement 2 for chunk1 and chunk2:
-					%% 1. chunk2 was previously read and cached
-					%% 2. chunk1 that was just read and used to compute H1
-					{noreply, update_sub_chunk_cache_size(-2, SessionKey, State3)}
-			end
+			ar_mining_stats:h1_solution(),
+			%% Update the cache to store h1.
+			case ar_mining_cache:with_cached_value(
+				?VDF_STEP_CACHE_KEY(Candidate#mining_candidate.cache_ref, Candidate#mining_candidate.nonce),
+				Candidate#mining_candidate.session_key,
+				State2#state.chunk_cache,
+				fun(CachedValue) -> {ok, CachedValue#ar_mining_cache_value{h1 = H1}} end
+			) of
+				{ok, ChunkCache1} -> State2#state{ chunk_cache = ChunkCache1 };
+				{error, Reason1} ->
```

Comment:
Is there any way we can hit `{error, Reason1}`? If not I'd say we remove the clause for now

Reply:
Ah I see, it's the `{error, cache_limit_exceeded}`

Previously we let the cache exceed the limit in order to avoid this sort of situation. I.e. by pre-allocating the cache before reading recall ranges we'll generally avoid the worst of the overruns. Occasionally we might overrun slightly, but never by more than 2 recall ranges (I don't think).

Do you have a sense of how often we'll hit the `cache_limit_exceeded` now? previously it was fairly often, if it's still fairly often then I suggest we go back to th eold behavir and allow the cache limit to be exceeded (otherwise the miner will likely see thousands of errors)

Reply:
This will conflict with the commitment to hold the cache size within the configured size.

Reply:
I mean, we need to formulate this commitment, because for now I don't understand what should we do with the cache size in different setups.

Reply:
Resolved in private conversation, the logic is to reserve the proper number of subchunks of space in the cache and delay the task just as we do with `compute_h0`.